sources/iTermPowerManager.m

```diff
@@ -11,29 +11,51 @@
 
 NSString *const iTermPowerManagerStateDidChange = @"iTermPowerManagerStateDidChange";
 
-@implementation iTermPowerManager
+@implementation iTermPowerManager {
+    CFRunLoopRef _runLoop;
+    CFRunLoopSourceRef _runLoopSource;
+}
+
+static void ITPMpowerSourceChanged(void *context) {
+    iTermPowerManager* pm = (__bridge iTermPowerManager *)(context);
+    
+    CFStringRef source = IOPSGetProvidingPowerSourceType(NULL);
+    if (source){
+        [pm setConnected:[@"AC Power" isEqualToString:(__bridge NSString *)(source)]];
+    }
+}
 
 + (instancetype)sharedInstance {
-    return [[self alloc] initPrivate];
+    static iTermPowerManager *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[self alloc] initPrivate];
+    });
+    
+    return sharedInstance;
 }
 
 - (instancetype)initPrivate {
     self = [super init];
     if (self) {
-        [self updateCharging];
-        [NSTimer it_scheduledTimerWithTimeInterval:5 repeats:YES block:^(NSTimer * _Nonnull timer) {
-            BOOL before = _connectedToPower;
-            [self updateCharging];
-            if (_connectedToPower != before) {
-                [[NSNotificationCenter defaultCenter] postNotificationName:iTermPowerManagerStateDidChange object:nil];
-            }
-        }];
+        _runLoop = CFRunLoopGetMain();
+        _runLoopSource = IOPSCreateLimitedPowerNotification(ITPMpowerSourceChanged,(__bridge void *)(self));
+        if (_runLoop && _runLoopSource){
+            CFRunLoopAddSource(_runLoop,_runLoopSource,kCFRunLoopDefaultMode);
+            dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
I think there need to be two changes here:

1. Eliminate dispatch_async so that the value of connectedToPower will be correct as soon as the object is initialized.
2. Don't call ITPMpowerSourceChanged because it causes a notification to be posted, and that doesn't make sense to do at this time. Callers aren't prepared to handle that notification at this time.

I think a better design would be to do:
```
static BOOL iTermPowerManagerIsConnectedToPower(void) {
    CFStringRef source = IOPSGetProvidingPowerSourceType(NULL);
    return [@"AC Power" isEqualToString:(__bridge NSString *)(source)]];
}

static void ITPMpowerSourceChanged(void *context) {
    iTermPowerManager* pm = (__bridge iTermPowerManager *)(context);
   [pm setConnected:iTermPowerManagerIsConnectedToPower()];
}
```

and in initPrivate, do:
```
_connectedToPower = iTermPowerManagerIsConnectedToPower();
```

Reply:
Thanks, looks perfect, I'll make the changes and squash everything in one commit.